CCE Conformal factor initial data

moxcodes opened the pull request:
## Proposed changes

Adds a CCE initial data generator based on choosing an angular conformal factor based on the value of the CCE scalar \beta. This is motivated by an attempt to make the time variable approximately asymptotically inertial.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

moxcodes committed:
Add ConformalFactor initial data generator

This generator uses the value of beta and J to choose a set of initial
hypersurface data such that the time coordinate is asymptotically inertial by
performing an iterative solve to find a set of angular coordinates in which the
asymptotic conformal factor is close to 1.

moxcodes committed:
Pass boundary Beta value to initial data generation

moxcodes committed:
Factor frequently-used block for initializing coordinates

moxcodes committed:
Add type aliases for often-used temp tags

moxcodes committed:
Pass node lock to CCE InitializeJ call operators

This is needed so that initialization schemes have the option of reading data
from file

moxcodes committed:
Add Swsh filtering overloads that accept a minimum and maximum l

Allows performing 'high-pass' filtering (e.g. excluding the l=0 mode)

moxcodes committed:
Factor iterative angular solve for re-use in conformal factor ID

moxcodes commented:
@keefemitman can you give this branch a try and make sure that it generates similarly good (or better) initial data to what you saw on the branch we were working with for a prototype?
The choice of options that should be most similar to our most recent iteration is probably something like:
```
ConformalFactor:
  AngularCoordinateTolerance: 1e-13
  MaxIterations: 1000
  RequireConvergence: True
  OptimizeL0Mode: True
  UseBetaIntegralEstimate: False
  ConfromalFactorIterationHeuristic: SpinWeight1CoordPerturbation
  UseInputModes: False
  InputModes: []
```

keefemitman commented:
> @keefemitman can you give this branch a try and make sure that it generates similarly good (or better) initial data to what you saw on the branch we were working with for a prototype?
> The choice of options that should be most similar to our most recent iteration is probably something like:
> 
> ```
> ConformalFactor:
>   AngularCoordinateTolerance: 1e-13
>   MaxIterations: 1000
>   RequireConvergence: True
>   OptimizeL0Mode: True
>   UseBetaIntegralEstimate: False
>   ConfromalFactorIterationHeuristic: SpinWeight1CoordPerturbation
>   UseInputModes: False
>   InputModes: []
> ```

Will do!

keefemitman commented on src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp:
```diff
@@ -185,8 +424,10 @@ struct InitializeJ<false> : public PUP::able {
+      gsl::not_null<Parallel::NodeLock*> hdf5_lock) const noexcept = 0;
 };
 }  // namespace InitializeJ
 }  // namespace Cce
```
Also need the line `#include "Evolution/Systems/Cce/Initialize/ConformalFactor.hpp"`

keefemitman commented on src/Evolution/Systems/Cce/Initialize/InitializeJ.hpp:
```diff
@@ -163,7 +401,8 @@ template <>
+                                   Tags::BoundaryValue<Tags::BondiBeta>>;
 
   using mutate_tags = tmpl::list<Tags::BondiJ, Tags::CauchyCartesianCoords,
                                  Tags::CauchyAngularCoords>;
```
Line 413 needs ConformalFactor as a part of the list

keefemitman reviewed (changes requested):
I was able to run the code with the new initial data scheme, but I found that I had to lower the angular tolerance to around 1e-4 instead of its suggested value of 1e-11. Is this ok?

moxcodes commented:
Thanks for checking that @keefemitman -- I would have anticipated that the algorithm would converge better than that. Can you try changing the OptimizeL0 option to `  OptimizeL0Mode: False`?
How does the value that you need to set for the threshold compare to the printed error measure from the prototype branch?

keefemitman commented:
@moxcodes With `OptimizeL0Mode` as false I still fail to obtain the specified tolerance. The print out is
"final maximum over collocation points deviation of J from target of 1.26732e-05"

moxcodes commented:
The high tolerance is concerning, but I think not necessarily a deal-breaker if the initial data still seems to be an improvement in terms of the transients we were seeing. Do you have any sense for how the waveform on this one compares with the waveforms you were seeing from the `cce_beta_based_id` branch?

keefemitman commented:
I'm still running a waveform right now, but I'll let you know once it finishes!

keefemitman commented:
Ok so it looks like the previous version of the conformal factor branch is much better than the new one in that the previous version of the branch largely reduced the transient effects at early times, but the newer version doesn't
<img width="436" alt="Screen Shot 2021-08-23 at 4 11 10 PM" src="https://user-images.githubusercontent.com/53911808/130530874-ffcc5fd6-c7cd-4dec-b67a-db5074ae11bb.png">

moxcodes commented:
Thanks, @keefemitman , I'll look into that and update this PR once I manage to find and fix the bug I must have introduced in the new version.

moxcodes commented:
I pushed a fixup and rebased. I believe that I've fixed the regression -- @keefemitman could you please test again and see if the behavior is now similar to the previous branch that has better initial transient results?
Thanks!

keefemitman commented:
@moxcodes Unfortunately I'm still seeing the same behavior as before, even with the new changes you made.

keefemitman commented:
> @moxcodes Unfortunately I'm still seeing the same behavior as before, even with the new changes you made.

Scratch that, this was just due to a different tolerance that I set in the input file. Everything looks good now!

kidder commented:
@nilsdeppe this has been rebased and sqaushed

nilsdeppe commented on src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp:
```diff
@@ -0,0 +1,219 @@
+                 ConformalFactorIterationHeuristic, UseInputModes,
+                 Options::Alternatives<tmpl::list<InputModesFromFile>,
+                                       tmpl::list<InputModes>>>;
+  static constexpr Options::String help = {""};
```
Maybe have some sort of non-empty string?

nilsdeppe commented on src/Evolution/Systems/Cce/Initialize/ConformalFactor.hpp:
```diff
@@ -0,0 +1,219 @@
+  static constexpr Options::String help = {""};
+
+  WRAPPED_PUPable_decl_template(ConformalFactor);  // NOLINT
+  explicit ConformalFactor(CkMigrateMessage* /*unused*/) {}
```
move definition to cpp and do `msg) : InitializeJ<false>(msg) {}`

nilsdeppe reviewed (changes requested):
LGTM, couple easy suggestions, rebase and squash right away please
